File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: None, val }
//     }
// }

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        // 1 -> 2 -> 3
        // node A = 1, node B = 2
        // Point B to A
        // Set A to B
        // Increment B

        if head.as_ref().is_none() || head.as_ref().unwrap().next.is_none() {
            return head;
        }
        let mut left = None;
        let mut right = head;

        while let Some(mut node) = right {
            let temp = node.next.take();
            node.next = left;
            left = Some(node);
            right = temp;
        }

        return left;
    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* sum = new ListNode();
        ListNode* curr = sum;
        
        int carry = 0;
        while(l1 && l2) {
            int v = l1->val + l2->val + carry;
            if(v > 9) {carry = 1;}
            else {carry = 0;}
            ListNode* next = new ListNode(v % 10);
            curr->next = next;
            curr = curr->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        while(l1) {
            int v = l1->val + carry;
            if(v > 9) {carry = 1;}
            else {carry = 0;}
            ListNode* next = new ListNode(v % 10);
            curr->next = next;
            curr = curr->next;
            l1 = l1->next;
        }
        while(l2) {
            int v = l2->val + carry;
            if(v > 9) {carry = 1;}
            else {carry = 0;}
            ListNode* next = new ListNode(v % 10);
            curr->next = next;
            curr = curr->next;
            l2 = l2->next;
        }
        if(carry > 0) {
            ListNode* next = new ListNode(carry);
            curr->next = next;
            curr = curr->next;
        }

        return sum->next;
    }
};
